File: src/year2021/month10/no322/CoinChange.java
package year2021.month10.no322;

import java.util.Arrays;

public class CoinChange {

    public static void main(String[] args) {
        int[] coins1 = {1, 2, 5};
        int amount1 = 11;
        int[] coins2 = {2};
        int amount2 = 3;
        int[] coins3 = {1};
        int amount3 = 0;
        int[] coins4 = {1};
        int amount4 = 1;
        int[] coins5 = {1};
        int amount5 = 2;
        System.out.println(coinChange(coins1, amount1));
        System.out.println(coinChange(coins2, amount2));
        System.out.println(coinChange(coins3, amount3));
        System.out.println(coinChange(coins4, amount4));
        System.out.println(coinChange(coins5, amount5));
    }

    public static int coinChange(int[] coins, int amount) {
        int[] dp = new int[amount + 1];
        final int MAX = amount + 1;
        Arrays.fill(dp, MAX);
        dp[0] = 0;
        Arrays.sort(coins);
        for (int i = 1; i <= amount; i++) {
            for (int coin : coins) {
                if (coin > i) {
                    break;
                }
                dp[i] = Math.min(dp[i], dp[i - coin] + 1);
            }
        }
        return dp[amount] == MAX ? -1 : dp[amount];
    }

}

/*
* 给你一个整数数组 coins ，表示不同面额的硬币；以及一个整数 amount ，表示总金额。

计算并返回可以凑成总金额所需的 最少的硬币个数 。如果没有任何一种硬币组合能组成总金额，返回 -1 。

你可以认为每种硬币的数量是无限的。

 

示例 1：

输入：coins = [1, 2, 5], amount = 11
输出：3 
解释：11 = 5 + 5 + 1
示例 2：

输入：coins = [2], amount = 3
输出：-1
示例 3：

输入：coins = [1], amount = 0
输出：0
示例 4：

输入：coins = [1], amount = 1
输出：1
示例 5：

输入：coins = [1], amount = 2
输出：2
 

提示：

1 <= coins.length <= 12
1 <= coins[i] <= 231 - 1
0 <= amount <= 104

来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/coin-change
著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
* */

File: go/year2025/month4/no322/CoinChange.go
package main

import "math"

func main() {
	coins1 := []int{1, 2, 5}
	amount1 := 11
	coins2 := []int{2}
	amount2 := 3
	coins3 := []int{1}
	amount3 := 0
	println(coinChange(coins1, amount1)) // 3
	println(coinChange(coins2, amount2)) // -1
	println(coinChange(coins3, amount3)) // 0
}

/*
DP, times: O(nk), space: O(n), n = amount, k = len(coins)
状态定义：dp[i] 表示组成金额为 i 的最少硬币数量
状态转移方程：dp[i] = min(dp[i-j] + 1), j <= i，j 为 coins 中可选的硬币
初始值：dp[0] = 0
*/
func coinChange(coins []int, amount int) int {
	dp := make([]int, amount+1)
	dp[0] = 0
	for i := 1; i <= amount; i++ {
		dp[i] = math.MaxInt
		for _, j := range coins {
			if i-j >= 0 && dp[i-j] != math.MaxInt {
				dp[i] = min(dp[i], dp[i-j]+1)
			}
		}
	}
	if dp[amount] == math.MaxInt {
		return -1
	}
	return dp[amount]
}

/*
给你一个整数数组 coins ，表示不同面额的硬币；以及一个整数 amount ，表示总金额。

计算并返回可以凑成总金额所需的 最少的硬币个数 。如果没有任何一种硬币组合能组成总金额，返回 -1 。

你可以认为每种硬币的数量是无限的。



示例 1：

输入：coins = [1, 2, 5], amount = 11
输出：3
解释：11 = 5 + 5 + 1
示例 2：

输入：coins = [2], amount = 3
输出：-1
示例 3：

输入：coins = [1], amount = 0
输出：0


提示：

1 <= coins.length <= 12
1 <= coins[i] <= 231 - 1
0 <= amount <= 104
*/
